Initial commit

Yngli opened the pull request:
The antlr4 grammar for the arden syntax version 2.10

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+exprComparison
+   : exprString
+   | exprFindString
+   | exprString singleCompOp exprString
+   | exprString IS NOT? mainCompOp
+   | exprString NOT? inCompOp
+   | exprString OCCUR NOT? temporalCompOp
+   | exprString OCCUR NOT? rangeCompOp
+   | exprString MATCHES PATTERN exprString
```
```suggestion
   : exprString
   | exprFindString
   | exprString singleCompOp exprString
   | exprString IS? NOT? mainCompOp
   | exprString OCCUR NOT? (temporalCompOp | rangeCompOp)
   | exprString MATCHES PATTERN exprString
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+exprString
+   : exprPlus
+   | exprString '||' exprPlus
+   | exprString FORMATTED WITH exprPlus
+   | TRIM trimOption exprString
+   | caseOption exprString
```
```suggestion
   : exprPlus
   | exprString ('||' | FORMATTED WITH) exprPlus
   | (TRIM trimOption | caseOption) exprString
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+
+exprPlus
+   : exprTimes
+   | exprPlus PLUS exprTimes
+   | exprPlus MINUS exprTimes
+   | PLUS exprTimes
+   | MINUS exprTimes
```
I beleive `PLUS` and `MINUS` have the same priority:

```suggestion
   | exprPlus (PLUS | MINUS) exprTimes
   | (PLUS | MINUS) exprTimes
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+exprAtTime
+   : exprBefore
+   | exprBefore ATTIME exprAtTime
```
```suggestion
   : exprBefore (ATTIME exprAtTime)?
```

Yngli replied:
Thank you for all the suggestions

KvanTTT commented on ardensyntax/ardensyntaxlexer.g4:
Please use [`caseInsensitive`](https://github.com/antlr/antlr4/blob/master/doc/options.md#caseinsensitive) option instead of multiple fragments like `[Aa]`, `[Bb]` and so on.

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+
+exprBefore
+   : exprAgo
+   | exprDuration BEFORE exprAgo
+   | exprDuration AFTER exprAgo
+   | exprDuration FROM exprAgo
```
```suggestion
   | exprDuration (BEFORE | AFTER | FROM) exprAgo
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+exprAgo
+   : exprDuration AGO?
+   | exprFunction AGO?
```
```suggestion
   : (exprDuration | exprFunction) AGO?
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+exprDuration
+   : exprFunction
+   | exprFunction durationOp
```
```suggestion
   : exprFunction durationOp?
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+exprSubListFrom
+   : SUBLIST exprFactor FROM exprFactor
+   | SUBLIST exprFactor STARTING AT exprFactor FROM exprFactor
```
```suggestion
   : SUBLIST exprFactor (FROM | STARTING AT exprFactor) exprFactor
```

KvanTTT replied:
Yes, to be precise, the following is more clear:

```
SUBLIST exprFactor (STARTING AT exprFactor)? FROM exprFactor
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   | NULL
+   | CONCLUDE
+   | IT
+   | LPAREN RPAREN
+   | LPAREN expr RPAREN
+   | LPAREN exprFuzzySet RPAREN
```
```suggestion
   | LPAREN (expr | exprFuzzySet)? RPAREN
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+booleanValue
+   : TRUE
+   | FALSE
+   | TRUTHVALUE NUMBER
+   | TRUTHVALUE TRUE
+   | TRUTHVALUE FALSE
```
```suggestion
   : TRUTHVALUE? (TRUE | FALSE)
   | TRUTHVALUE NUMBER
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+dataAssignPhrase
+   : READ readPhrase
+   | READ AS IDENTIFIER readPhrase
+   | MLM TERM
+   | MLM TERM FROM INSTITUTIONWC string
+   | MLM MLM_SELF
+   | INTERFACE mappingFactor
+   | EVENT mappingFactor
+   | MESSAGE mappingFactor
+   | MESSAGE AS IDENTIFIER mappingFactor?
+   | DESTINATION mappingFactor
+   | DESTINATION AS IDENTIFIER mappingFactor?
```
```suggestion
   : READ (AS IDENTIFIER)? readPhrase
   | MLM (TERM (FROM INSTITUTIONWC string)? | MLM_SELF)
   | (INTERFACE | EVENT | MESSAGE) mappingFactor
   | MESSAGE AS IDENTIFIER mappingFactor?
   | DESTINATION (mappingFactor | AS IDENTIFIER mappingFactor?)
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+objectAttributeList
+   : IDENTIFIER
+   | IDENTIFIER COMMA objectAttributeList
```
```suggestion
   : IDENTIFIER (COMMA objectAttributeList)?
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+newObjectPhrase
+   : NEW IDENTIFIER (WITH expr)?
+   | NEW IDENTIFIER WITH LBRACK objectInitList RBRACK
+   | NEW IDENTIFIER WITH expr WITH LBRACK objectInitList RBRACK
```
```suggestion
   : NEW IDENTIFIER (WITH (expr | (expr WITH)? LBRACK objectInitList RBRACK))?
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+evokeStatement
+   : (eventOr)?
```
```suggestion
   : eventOr?
```

KvanTTT commented on ardensyntax/ardensyntax.g4:
```diff
@@ -0,0 +1,919 @@
+   ;
+
+eventAny
+   : ANY OF? LPAREN eventList RPAREN
+   | ANY OF? IDENTIFIER
```
```suggestion
   : ANY OF? (LPAREN eventList RPAREN | IDENTIFIER)
```

KvanTTT reviewed (changes requested):
Thanks for the contribution!

But please fix the following issues:

1. Use [caseInsensitive](https://github.com/antlr/antlr4/blob/master/doc/options.md#caseinsensitive) lexer option to get rid of fragmetns. It makes grammar more readable.
2. Factor common parts in rules if possible in parser, use `?` operator.

kaby76 commented:
Please make the following changes.
* Rename ardensyntax.g4 to ArdenParser.g4 -- The Antlr tool does not like the file name to be different from the grammar name.
* Rename ardensyntaxlexer.g4 to ArdenLexer.g4 -- ditto. In addition, the tester will not work.
* Rename the directory ardensyntax/ to just arden/ -- all grammars describe a syntax. It's understood and not necessary to belabor the point.
* Move examples/Examples.md to ".." or just remove.
* examples/Fifth.mlm fail 11.8614948 -- the example does not parse.
* The grammar is exceedingly slow. As it stands, it's not clear if any targets other than Java may complete. The desc.xml will need to be modified. But, it would be best to understand why the grammar is slow, using performance tracing. Alternatively, you could partition the tests into fast/ and slow/ directories, then have a desc.xml point to both depending on the target. Here is the runtime for a CSharp build.
```
$ ./bin/Debug/net7.0/Test.exe ../examples/*
line 16:0 mismatched input 'http' expecting ';;'
CSharp 0 ../examples/Fifth.mlm fail 11.8614948
CSharp 1 ../examples/First.mlm success 17.7846296
CSharp 2 ../examples/Fourth.mlm success 120.0488035
CSharp 3 ../examples/Second.mlm success 20.629743
CSharp 4 ../examples/Third.mlm success 56.4346655
Total Time: 226.8077866
06/12-09:47:51 ~/arden/ardensyntax/Generated-CSharp
```

Yngli commented on ardensyntax/ardensyntax.g4:
I've renamed both like suggested from kaby76

KvanTTT commented on arden/ArdenLexer.g4:
```diff
@@ -0,0 +1,344 @@
+ARCSIN:         'arcsin';
+ARCTAN:         'arctan';
+ARDEN:          'arden:';
+ARDEN_VERSION:  'version'[ ]('2' (DOT ( '1' | '2' | '5' | '6' | '7' | '8' | '9' | '10'))? | [3]);
```
Please don't mix string literals (`' '`) and range literals (`[ ]`) within one rule:

```suggestion
ARDEN_VERSION:  'version ' ('2' ('.' ( '1' | '2' | '5' | '6' | '7' | '8' | '9' | '10'))? | '3');
```

KvanTTT commented on arden/ArdenLexer.g4:
```diff
@@ -0,0 +1,344 @@
+
+TERM:           '\''  .*? '\'';
+
+STRING:         '"' ( '""' | ~["] )* '"';
```
```suggestion
STRING:         '"' ( '""' | ~'"' )* '"';
```

Yngli commented:
Thank you for all suggestions.
I hope i provided all necessary fixes.

@kaby76 The speed of the grammar is now better after these changes. 
The grammar has still some ambiguities. I will speak with workgroup of arden, to remove them. This will hopefully increase the performance even more.

Thank you for the coverage. It's probably even more. I have way more tests for it but I can't share them here, because it's code from products of the company.

KvanTTT commented on arden/ArdenLexer.g4:
```diff
@@ -0,0 +1,344 @@
+
+// Start of the identifier 10 chars, starting with a letter
+fragment StartIdentifier
+    :           Letter ID? ID? ID? ID? ID? ID? ID? ID? ID?
```
Did you mean the following? It accepts a sequence from 0 to 9 `ID` elements:

```suggestion
    :           Letter (ID (ID (ID (ID (ID (ID (ID (ID ID?)?)?)?)?)?)?)?)?
```

The same applicable to other similar rules defined below (`RestIdentifier`, `MlmIDStart`, `MlmIDRest`).

Yngli replied:
Yes this is exactly what I mean.

KvanTTT commented on arden/ArdenLexer.g4:
```diff
@@ -0,0 +1,344 @@
+TEXT:           .+? ';;' -> mode(DEFAULT_MODE);
+
+mode DataMapping;
+DATA_MAPPING:   (~[}] | '\\' .)+ -> mode(DEFAULT_MODE);
```
```suggestion
DATA_MAPPING:   (~'}' | '\\' .)+ -> mode(DEFAULT_MODE);
```

KvanTTT commented on arden/ArdenLexer.g4:
```diff
@@ -0,0 +1,344 @@
+TWOCHARCODE    options {caseInsensitive = false; }: Letter Letter ('_' [A-Z] [A-Z])? -> mode(DEFAULT_MODE);
+
+WS_TCM:         WS -> channel(HIDDEN);
+
+
+
+
```
Please remove excess line breaks:

```suggestion
```

kaby76 commented:
Looks like there are other issues.

* Mutual left-recursion. Rewrite using Kleene `*`-operator.
```
error(119): ArdenParser.g4::: The following sets of rules are mutually left-recursive [languageSlot]

languageSlot
   : languageSlot? singleLanguageCode
   ;

=>

languageSlot
    : singleLanguageCode+
    ;
```
* Rewrite start rule so that the build tools can find an EOF-terminated rule. Rewrite mlms using Kleene operator.
```
mlms
   : mlms mlm
   | mlm
   | EOF
   ;

=>
file_
    : mlns EOF
    ;

mlms
    : mlm*
    ;
```
* Errors in parsing all files, e.g., First.mlm:
```
First.mlm.errors
line 53:17 no viable alternative at input '"The calculated fractional excretion of sodium is low ("\n|| fractional_na ||'
line 64:12 no viable alternative at input '"The calculated fractional excretion of sodium is " ||\n"not low (" ||'
Fourth.mlm.errors
line 71:27 no viable alternative at input '"Increased dose of beta blockers may be " ||\n"needed to control angina.";'
line 76:29 no viable alternative at input '"Increased dose of beta blockers may " ||\n"be needed to control PVC's.";'
Second.mlm.errors
line 61:31 no viable alternative at input '"calcium = " || calcium ||'
line 65:26 no viable alternative at input 'msg||"; albumin = "||'
line 68:22 no viable alternative at input 'msg ||\n"; last creatinine = "||'
line 70:33 no viable alternative at input 'msg ||\n"; (total or corrected calcium " ||'
line 81:31 no viable alternative at input '"calcium = " || calcium ||'
line 83:32 no viable alternative at input 'msg || "; last albumin " ||'
line 88:0 no viable alternative at input 'msg || "; last creatinine = "\n||'
line 90:39 no viable alternative at input 'msg ||\n"; (total calcium was at least 11.0; " ||'
line 100:44 no viable alternative at input '"hypercalcemia study: " || msg;;'
line 140:21 no viable alternative at input '"Caution, the patient has the following allergy to penicillin documented:"\n|| penicillin_allergy;;'
Third.mlm.errors
line 64:48 no viable alternative at input '"Due to renal insufficiency, the dose of gentamicin " ||\n"should be adjusted. The patient's calculated " ||'
line 129:49 no viable alternative at input '"Suggest obtaining a serum creatinine to follow up " ||\n"on renal function in the setting of gentamicin.";'
line 131:55 no viable alternative at input '"Recent serum creatinine (" || recent_creatinine ||'
line 188:47 no viable alternative at input '"Caution: patient's relative granulocytopenia may be " ||\n"exacerbated by trimethoprim/sulfamethoxazole.";'
Test failed.
```
* There are references to "turtle" grammar in the pom.xml.

kaby76 commented:
* Things are still not parsing. `exprString` seems to have a typo.
```
exprString
   : exprPlus
   | exprString ('||' exprPlus | FORMATTED WITH) exprPlus

...

=>

exprString
   : exprPlus
   | exprString ('||' | FORMATTED WITH) exprPlus
...
```
* The grammar is too slow for JavaScript, TypeScript, Python3, and PHP. And something is wrong with the Go target code with this grammar--it hangs on Fourth.mlm. Please remove these targets from the desc.xml. In addition, the grammar uses `options { caseInsensitive = true; }`. So, the desc.xml should indicate that the grammar cannot be used on any version of Antlr before 4.10.
```
<?xml version="1.0" encoding="UTF-8" ?>
<desc xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../_scripts/desc.xsd">
   <antlr-version>^4.10</antlr-version>
    <targets>CSharp;Cpp;Dart;Java</targets>
</desc>
```


```

Yngli commented:
I've added all suggestions, thank you again for all these improvements.
I've removed the other targets, unfortunately the gramar is in this state very slow.
I will try in the upcoming versions to improve the performance of the grammar.

Yngli commented:
Yes I know, there are unfortnately a couple more.
I'm currently trying to prove, if I remove all this ambiguties, that all statemtents still work fine.
I'm pretty sure with the ambiguity you found, to just delete exprFunction and it's fine.

teverett commented:
@Yngli  thanks!
